### PR TITLE
Fix intsimdmatrixneon.cpp stack corruption.

### PR DESCRIPTION
**Pull request by robinwatts**
My understanding of the intsimdmatrix mechanism was that the
inputs would be resized so that we'd only ever get "whole
blocks" of data (i.e. that num_out would be a multiple of
kNumOutputsPerRegister, and num_in would be a multiple of
kNumInputsPerRegister). Sadly, this does not seem to be the
case, as we can get called (sometimes) with num_out % 8 == 7.

It's not clear to me WHY this happens. Is it a mistake in
the calling code? A quick investigation reveals that the
offending matrices are being read from the file and are not
obviously resized before being used.

Possibly we could benefit from resizing those matrices so
that special cases in this innermost loop are not actually
required, but unless and until that is done, let's fix the
inner loop.

### CODE REVIEW

**Comment by robinwatts** [comment]
I don't believe that the failing test is the fault of my commit.

**Comment by stweil** [comment]
My test now passes without problems, thanks.

File: src/arch/intsimdmatrixneon.cpp [hunk 1]
```diff
@@ -126,19 +137,19 @@ static void matrixDotVector(int dim1, int dim2, const int8_t* wi,
...
-  for (; output + group_size <= rounded_num_out; output += group_size) {
-    PartialMatrixDotVector8(wi, scales, u, rounded_num_in, v);
+  for (; output + group_size <= num_out; output += group_size) {
+    PartialMatrixDotVector8(wi, scales, u, rounded_num_in, v, 8);
     wi += w_step;
     scales += group_size;
     v += group_size;
   }
+  if (output < num_out)
+    PartialMatrixDotVector8(wi, scales, u, rounded_num_in, v, num_out & 7);
```
[the remark this answers is not in the document]
Comment (robinwatts) [reply src/arch/intsimdmatrixneon.cpp:152]: Moving from & to % is bad.

num_out is signed, so (as far as the compiler knows) might be negative.

x & 7 != x % 8 when x might be negative. Therefore the compiler can't do a simple "and" operation; it probably ends up calling a helper function, which is not ideal in the single most speed-critical region of the code.

I'll use "& (kNumOutputsPerRegister-1)".

[the remark this answers is not in the document]
Comment (stweil) [reply src/arch/intsimdmatrixneon.cpp]: That's fine, too.

**Pull request merged**